chia/full_node/full_node.py

```diff
@@ -2351,26 +2343,71 @@ async def add_transaction(
-                assert cost is not None
-                new_tx = full_node_protocol.NewTransaction(
-                    spend_name,
-                    cost,
-                    fees,
-                )
-                msg = make_msg(ProtocolMessageTypes.new_transaction, new_tx)
-                if peer is None:
-                    await self.server.send_to_all([msg], NodeType.FULL_NODE)
-                else:
-                    await self.server.send_to_all([msg], NodeType.FULL_NODE, peer.peer_node_id)
+
+                await self.notify_peers_of_mempool_item(mempool_item, peer)
+
                 if self.simulator_transaction_callback is not None:  # callback
                     await self.simulator_transaction_callback(spend_name)  # pylint: disable=E1102
+
             else:
                 self.mempool_manager.remove_seen(spend_name)
                 self.log.debug(f"Wasn't able to add transaction with id {spend_name}, status {status} error: {error}")
         return status, error
 
+    async def notify_peers_of_mempool_item(
+        self, mempool_item: MempoolItem, peer: Optional[WSChiaConnection] = None
+    ) -> None:
+        new_tx = full_node_protocol.NewTransaction(
+            mempool_item.spend_bundle_name,
+            mempool_item.cost,
+            mempool_item.fee,
+        )
+        new_tx_msg = make_msg(ProtocolMessageTypes.new_transaction, new_tx)
+
+        await self.server.send_to_all([new_tx_msg], NodeType.FULL_NODE, None if peer is None else peer.peer_node_id)
+
+        wallet_update = wallet_protocol.TransactionAddedUpdate(
+            mempool_item.spend_bundle_name,
+            mempool_item.height_added_to_mempool,
+            mempool_item.cost,
+            mempool_item.additions,
+            mempool_item.removals,
+        )
+        wallet_update_msg = make_msg(ProtocolMessageTypes.transaction_added_update, wallet_update)
+
+        peers = await self.peers_for_mempool_item(mempool_item)
+
+        def new_clients(conn: WSChiaConnection) -> bool:
+            return conn.peer_node_id in peers and conn.protocol_version > Version("0.0.36")
+
+        await self.server.send_to_all_if(
+            [wallet_update_msg], NodeType.WALLET, new_clients, None if peer is None else peer.peer_node_id
+        )
+
+    async def peers_for_mempool_item(self, mempool_item: MempoolItem) -> Set[bytes32]:
+        all_coins = mempool_item.additions + mempool_item.removals
+
+        peers: Set[bytes32] = set()
+        coin_ids: List[bytes32] = []
+
+        for coin in all_coins:
+            peers_for_ph = self.subscriptions.peers_for_puzzle_hash(coin.puzzle_hash)
+            peers.update(peers_for_ph)
+
+            coin_id = coin.name()
+            peers_for_coin = self.subscriptions.peers_for_coin_id(coin_id)
+            peers.update(peers_for_coin)
+
+            coin_ids.append(coin_id)
+
+        hints = [bytes32(hint) for hint in await self.hint_store.get_hints_for_coins(coin_ids) if len(hint) == 32]
```

Comment:
this is an expensive operation. Are there ways to avoid these lookups unconditionally, every time the mempool is updated?

Reply:
This would be for e.g. detecting that one of your NFTs have been spent or created. I don't know of a faster way to do this, at least it's one query

Reply:
Moving to another PR later.